src/natural_history_manager.rs

```diff
@@ -0,0 +1,572 @@
+        let times = &natural_history_container.times[gi_index];
+        let gi_trajectory = &natural_history_container.gi_trajectories[gi_index];
+        // Because we want to interpolate the *inverse* CDF, the CDF values are "x" and the time values are "y".
+        interpolate(gi_trajectory, times, gi_cdf_value_unif)
+    }
+}
+
+/// An interpolation routine that expects a paired set of values `xs` and `ys` that represent samples
+/// from a given function. The function is evaluated at a given x value `xp` using cubic spline interpolation
+/// when there are at least two samples above and below `xp`. Otherwise, it uses linear extrapolation at the tails.
+/// Assumes that function samples are sorted so that the `xs` are in ascending order.
+fn interpolate(xs: &[f64], ys: &[f64], xp: f64) -> f64 {
+    let upper_window_index_option = xs.iter().position(|&x| x > xp);
+    // We need to check whether a point was found. If it wasn't, it means that all values
+    // in `xs` are less than `xp`. We have to use an alternative extrapolation strategy.
+    let upper_window_index = match upper_window_index_option {
+        None => {
+            // We are above the range of the `xs` samples, so we must extrapolate. We default to
+            // linear extrapolation using the last two values in `xs`.
+            let traj_len = xs.len();
+            return linear_interpolation(
+                xs[traj_len - 2],
+                xs[traj_len - 1],
+                ys[traj_len - 2],
+                ys[traj_len - 1],
+                xp,
+            );
+        }
+        Some(i) => match i {
+            // We only have one point in `xs` below `xp`, so we must also default to linear extrapolation.
+            1 => return linear_interpolation(xs[0], xs[1], ys[0], ys[1], xp),
+            // Else, we use cubic spline interpolation.
+            // The index can never return 0 because we handle that case at the beginning of the function.
+            _ => i,
+        },
+    };
+    // If our interpolation point is in the range of the `xs` samples, we can use cubic spline interpolation.
+    cubic_spline_interpolation(
+        xs[(upper_window_index - 2)..=(upper_window_index + 1)]
+            .try_into()
+            .unwrap(),
+        // Since `upper_window_index` is the third index in the trajectory, we need to subtract 2 to get the first index.
+        ys[(upper_window_index - 2)..=(upper_window_index + 1)]
+            .try_into()
+            .unwrap(),
+        xp,
+    )
+}
+
+/// Fits a line between two points and returns the value of the line at a given x-value `xp`.
+/// Sensibly handles equal x-values by returning `y1` for `xp < x1`, `y2` for `xp > x2`, and
+/// `(y1 + y2) / 2` for `xp = x1 = x2`.
+fn linear_interpolation(x1: f64, x2: f64, y1: f64, y2: f64, xp: f64) -> f64 {
+    // At the tails of the CDF, the CDF moves very slowly, so with numerical imprecision, x2 may equal x1.
+    #[allow(clippy::float_cmp)]
+    if x2 == x1 {
```

Comment:
I'm trying to figure out how you get into this state. By construction, isn't `xp < x2`?